File: lib/domain/entities/investment_totals.dart
import 'package:decimal/decimal.dart';

import '../services/portfolio_analytics.dart';
import 'asset_group.dart';

/// Everything the rest of the app needs to know about the portfolio.
///
/// This is the single seam between the Investments feature and every other
/// consumer of investment value: the Dashboard's tile, the health score, the
/// safety net, account net worth, and the daily snapshot.
///
/// Deliberately **not** [Position] or [PortfolioSnapshot]. Two reasons:
///
///  * The health score must not depend on FIFO lot matching. Coupling a score
///    to the disposal algorithm means every change to one is a change to both.
///  * The web client builds this from its own model, so the type has to be
///    expressible without Dart-side lot machinery.
///
/// Before this existed, each consumer summed the legacy `Holdings` table itself
/// while the Investments screen read the lot model — so the Dashboard and the
/// Investments screen showed different numbers for the same portfolio, and
/// sample data (which only wrote the legacy table) rendered a large net worth
/// beside an empty Investments screen. One type, one derivation, one number.
class InvestmentTotals {
  const InvestmentTotals({
    required this.marketValue,
    required this.costBasis,
    required this.valueByGroup,
    required this.positionCount,
    required this.unpricedCount,
    required this.indicativeValue,
    this.lastPricedAt,
    this.unconvertedCurrencies = const {},
  });

  /// What the portfolio is worth. Unpriced positions contribute their cost
  /// basis, so this never silently drops a holding — see [unpricedCount] and
  /// [indicativeValue] for how much of it is not a real market price.
  final Decimal marketValue;

  /// What was paid, charges included.
  final Decimal costBasis;

  /// Market value per chart group. Every present group has an entry; absent
  /// groups are simply missing rather than mapped to zero.
  final Map<AssetGroup, Decimal> valueByGroup;

  /// Distinct instruments currently held. Zero means the vault genuinely has no
  /// investments — which callers must keep distinct from "not loaded yet".
  final int positionCount;

  /// Positions with no recorded price at all.
  final int unpricedCount;

  /// How much of [marketValue] rests on something other than a market quote:
  /// positions carried at cost, priced by hand, or carrying a price whose date
  /// is not real. A total that is a third guesswork should not be presented
  /// with the same confidence as one that isn't.
  final Decimal indicativeValue;

  /// The most recent price observation anywhere in the portfolio.
  final DateTime? lastPricedAt;

  /// Currencies held that could not be converted, because no rate is stored.
  ///
  /// Those holdings are **excluded** from [marketValue] rather than counted at
  /// parity. Counting 100 USD as 100 INR would understate net worth by 99% and
  /// look entirely plausible on screen; an obviously missing number sends the
  /// user to Settings, a quietly wrong one does not.
  final Set<String> unconvertedCurrencies;

  static final empty = InvestmentTotals(
    marketValue: Decimal.zero,
    costBasis: Decimal.zero,
    valueByGroup: const {},
    positionCount: 0,
    unpricedCount: 0,
    indicativeValue: Decimal.zero,
  );

  /// True when some holding could not be expressed in the base currency.
  bool get hasUnconverted => unconvertedCurrencies.isNotEmpty;

  /// True when the vault holds nothing.
  ///
  /// Distinct from "still loading" on purpose: callers must represent that as
  /// an absent value, never as [empty]. Treating a loading portfolio as an
  /// empty one makes the health score visibly dip on every cold open.
  bool get isEmpty => positionCount == 0;

  /// Unrealised gain or loss.
  Decimal get unrealisedPnl => marketValue - costBasis;

  /// Value of the retirement-bucket assets (FD / PPF-EPF / NPS).
  Decimal get retirementValue =>
      valueByGroup[AssetGroup.retirement] ?? Decimal.zero;

  /// Value in growth assets — equity and gold. Used by the score's Future
  /// category, where "am I invested for growth" is a different question from
  /// "am I invested at all".
  Decimal get growthValue =>
      (valueByGroup[AssetGroup.equity] ?? Decimal.zero) +
      (valueByGroup[AssetGroup.gold] ?? Decimal.zero);

  /// The largest single group's share of the portfolio, 0..1. Null when there
  /// is nothing held — an undefined concentration must not read as zero.
  double? get concentration {
    if (marketValue <= Decimal.zero || valueByGroup.isEmpty) return null;
    final largest = valueByGroup.values
        .fold(Decimal.zero, (a, b) => b > a ? b : a);
    return (largest / marketValue).toDouble();
  }

  /// Builds the totals, converting any foreign holding into [baseCurrency].
  ///
  /// [rateFor] returns base units per 1 unit of the given currency, or null
  /// when no rate is known. Passing null for [rateFor] keeps every amount at
  /// face value — correct only for a single-currency vault, which is why the
  /// providers always supply one.
  factory InvestmentTotals.fromSnapshot(
    PortfolioSnapshot snap, {
    String baseCurrency = 'INR',
    Decimal? Function(String currency)? rateFor,
  }) {
    final byGroup = <AssetGroup, Decimal>{};
    final unconverted = <String>{};
    var market = Decimal.zero;
    var cost = Decimal.zero;
    var indicative = Decimal.zero;
    var counted = 0;

    for (final p in snap.positions) {
      final currency = p.instrument.currency;
      Decimal rate = Decimal.one;
      if (currency != baseCurrency) {
        final found = rateFor?.call(currency);
        if (found == null) {
          // Excluded, and named, rather than counted at parity.
          unconverted.add(currency);
          continue;
        }
        rate = found;
      }

      final value = p.marketValue * rate;
      final g = p.instrument.group;
      byGroup[g] = (byGroup[g] ?? Decimal.zero) + value;
      market += value;
      cost += p.costBasis * rate;
      counted++;
      // Not just unpriced: a bond you typed a price for last month is no more
      // a market quote than one you never priced at all.
      if (p.isIndicative) indicative += value;
    }

    return InvestmentTotals(
      marketValue: market,
      costBasis: cost,
      valueByGroup: byGroup,
      positionCount: counted,
      unpricedCount: snap.unpriced.length,
      indicativeValue: indicative,
      lastPricedAt: snap.lastPricedAt,
      unconvertedCurrencies: unconverted,
    );
  }
}

File: webapp/src/domain/investmentTotals.ts
// The single seam between the portfolio and everything that consumes its value.
//
// Mirrors `lib/domain/entities/investment_totals.dart`. Keep the two in step:
// the health score, the safety net and the net-worth snapshot all take this
// type on both platforms, and the shared fixture in health.test.ts asserts they
// produce identical numbers from it.
import Decimal from 'decimal.js';
import { D, ZERO } from '@/lib/money';
import type { Holding } from '@/lib/types';
import { ASSET_GROUP_OF, holdingView, type AssetGroup } from './portfolio';

/**
 * Everything outside the Investments screen needs to know about the portfolio.
 *
 * Deliberately not a holdings list: consumers should not each re-derive "what
 * is this worth", because when they did, they disagreed. Deliberately not the
 * position/lot type either — the health score must not depend on how a lot is
 * matched, and this type has to be expressible on both clients.
 *
 * Model note, and it is a real difference rather than an oversight: Flutter
 * builds this from the lot ledger (instruments + trades + dated prices), while
 * the web app builds it from aggregated holdings (one row per symbol with an
 * average cost). Both produce the same market value and cost basis for the same
 * portfolio. What web cannot produce is realised P&L or a price date, because
 * an aggregate row has neither. Anything that needs those must read the lot
 * model directly and is Flutter-only for now.
 */
export interface InvestmentTotals {
  /**
   * What the portfolio is worth. Unpriced holdings contribute their cost, so
   * this never silently drops one — see `unpricedCount` and `indicativeValue`.
   */
  marketValue: Decimal;
  /** What was paid. */
  costBasis: Decimal;
  /** Market value per chart group. Absent groups are missing, not zero. */
  valueByGroup: Partial<Record<AssetGroup, Decimal>>;
  /** Distinct holdings. Zero means the vault holds nothing. */
  positionCount: number;
  /** Holdings with no recorded price. */
  unpricedCount: number;
  /** How much of `marketValue` is carried at cost rather than a real quote. */
  indicativeValue: Decimal;
  /**
   * The newest price observation across the priced holdings, or null when
   * nothing carries a date.
   *
   * Was declared and hard-coded null at every construction site, so any screen
   * that trusted it showed nothing. It is real now that holdings record
   * `priceAsOf`, and it is what lets a total say how old it is.
   */
  lastPricedAt: number | null;
}

export const EMPTY_INVESTMENTS: InvestmentTotals = {
  marketValue: ZERO,
  costBasis: ZERO,
  valueByGroup: {},
  positionCount: 0,
  unpricedCount: 0,
  indicativeValue: ZERO,
  lastPricedAt: null,
};

/** True when the vault holds nothing. Distinct from "not loaded yet". */
export function isEmptyPortfolio(t: InvestmentTotals): boolean {
  return t.positionCount === 0;
}

/** Value of the retirement bucket (FD / PPF-EPF / NPS). */
export function retirementValue(t: InvestmentTotals): Decimal {
  return t.valueByGroup.retirement ?? ZERO;
}

/** Equity + gold — "invested for growth", a different question from "invested". */
export function growthValue(t: InvestmentTotals): Decimal {
  return (t.valueByGroup.equity ?? ZERO).plus(t.valueByGroup.gold ?? ZERO);
}

/**
 * The largest group's share, 0..1. Null when nothing is held — an undefined
 * concentration must not read as zero, which would score as perfect diversity.
 */
export function concentration(t: InvestmentTotals): number | null {
  if (t.marketValue.lte(0)) return null;
  const values = Object.values(t.valueByGroup) as Decimal[];
  if (values.length === 0) return null;
  const largest = values.reduce((a, b) => (b.gt(a) ? b : a), ZERO);
  return largest.div(t.marketValue).toNumber();
}

/** Builds the totals from the web app's aggregated holdings. */
export function investmentTotals(holdings: Holding[]): InvestmentTotals {
  const byGroup: Partial<Record<AssetGroup, Decimal>> = {};
  let market = ZERO;
  let cost = ZERO;
  let indicative = ZERO;
  let unpriced = 0;
  let newest: number | null = null;

  for (const h of holdings) {
    const v = holdingView(h);
    const g = ASSET_GROUP_OF[h.assetType];
    byGroup[g] = (byGroup[g] ?? ZERO).plus(v.current);
    market = market.plus(v.current);
    cost = cost.plus(v.invested);
    if (h.lastPrice == null || h.lastPrice === '') {
      unpriced += 1;
      indicative = indicative.plus(v.current);
    } else if (h.priceAsOf != null && (newest == null || h.priceAsOf > newest)) {
      newest = h.priceAsOf;
    }
  }

  return {
    marketValue: market,
    costBasis: cost,
    valueByGroup: byGroup,
    positionCount: holdings.length,
    unpricedCount: unpriced,
    indicativeValue: indicative,
    lastPricedAt: newest,
  };
}

/** Convenience for tests and fixtures: totals from plain per-asset-type amounts. */
export function totalsOf(
  byType: Partial<Record<import('@/lib/types').AssetType, number>>,
  opts: { costBasis?: number; unpricedCount?: number; indicativeValue?: number } = {},
): InvestmentTotals {
  const byGroup: Partial<Record<AssetGroup, Decimal>> = {};
  let total = ZERO;
  let count = 0;
  for (const [type, amount] of Object.entries(byType)) {
    if (amount == null) continue;
    const g = ASSET_GROUP_OF[type as import('@/lib/types').AssetType];
    const v = D(amount);
    byGroup[g] = (byGroup[g] ?? ZERO).plus(v);
    total = total.plus(v);
    count += 1;
  }
  return {
    marketValue: total,
    costBasis: opts.costBasis == null ? total : D(opts.costBasis),
    valueByGroup: byGroup,
    positionCount: count,
    unpricedCount: opts.unpricedCount ?? 0,
    indicativeValue: D(opts.indicativeValue ?? 0),
    lastPricedAt: null,
  };
}
